### PR TITLE
Added matrix-type independent add in linalg

### PR DESCRIPTION
**Pull request by lambday**
@karlnapf @yorkerlin

### CODE REVIEW

File: src/shogun/mathematics/linalg/internal/implementation/Add.h [hunk 1]
```diff
@@ -96,10 +119,12 @@ struct add<Backend::NATIVE, Matrix>
 	static void compute(SGMatrix<T> A, SGMatrix<T> B, SGMatrix<T> C,
 		T alpha=1, T beta=1)
 	{
-		REQUIRE(A.num_rows == B.num_rows && B.num_rows == C.num_rows,
-			"Matrices should have same number of rows!\n");
-		REQUIRE(A.num_cols == B.num_cols && B.num_cols == C.num_cols,
-			"Matrices should have same number of columns!\n");
+		REQUIRE(A.matrix, "Matrix A is not initialized!\n");
+		REQUIRE(B.matrix, "Matrix B is not initialized!\n");
```
Comment (lambday) [inline src/shogun/mathematics/linalg/internal/implementation/Add.h]: Missed a few REQUIRES. Adding.

**Comment by lambday** [comment]
Travis failure unrelated. Merging.

**Comment by karlnapf** [comment]
+1

**Comment by iglesias** [comment]
Hello, @lambday. Do you think it would make sense / be worth to use inheritance (or something else, maybe) in order to avoid some code duplication here?

For instance, we have now the specialised structs `add<Backend::Native, Matrix>`, `add<Backend::EIGEN3, Matrix>` , etc. with the same method signature and implementation for
`static SGMatrix<T> compute(SGMatrix<T> A, SGMatrix<T> B, T alpha=1, T beta=1) `,
among other methods.

**Comment by lambday** [comment]
@iglesias nice point. As of the now, the implementation (apart from the REQUIRES) is (just a tiny bit) different. For this method,
```c++
static SGMatrix<T> compute(SGMatrix<T> A, SGMatrix<T> B, T alpha=1, T beta=1)
```
with NATIVE, calls a `compute()` that handles raw pointers, 
```c++
static void compute(T* A, T* B, T*C, T alpha, T beta, index_t len)
```
whether the EIGEN3 one calls a different `compute()`.
```c++
static void compute(SGMatrix<T> A, SGMatrix<T> B, SGMatrix<T> C, T alpha=1, T beta=1)
```
The inheritance thing can be useful if we make the NATIVE one call the same `compute()` that we're using for Eigen3. I used the raw-pointer one for NATIVE to avoid additional method calls (because the other one in turn calls that very same method in NATIVE - so one additional method call if we do that).

**Comment by iglesias** [comment]
You are right, I didn't notice the difference in the types of the arguments. Thanks for explaining ;-)

**Pull request merged**